File: part_2/svr.R
# :help Nvim-R
# \rf - start r session
# \rq - close r session
# \l  — send current line
# \ae - Send file
# \se — send current selection (and echo)

library(caTools)
library(ggplot2)
library(e1071)

dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Fitting Polynomial Regression Model to the dataset
# For example:
regressor = svm(formula = Salary ~ .,
                data = dataset,
                type = 'eps-regression')

# PREDICTION
# For example:
yPred1 = predict(regressor, data.frame(Level = 6.5))
# yPred2 = predict(regressor, data.frame(Level = 6.5,
#                                        Level2 = 6.5^2,
#                                        Level3 = 6.5^3,
#                                        Level4 = 6.5^4))

# VISUALIZATION
# Polynomial Regression
# For Example:
ggplot() +
    geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = "red") +
    geom_line(aes(x = dataset$Level, y = predict(regressor, newData = dataset)), colour = "blue") +
    ggtitle("Truth or Bluff (Linear Regression)") +
    xlab("position level") +
    ylab("salary")

# polynomial with higher resolution
xGrid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
    geom_point(aes(x = dataset$Level ,y = dataset$Salary), colour="red") +
    geom_line(aes(x = xGrid, y = predict(regressor,
                                         newdata = data.frame(Level = xGrid,
                                                              Level2 = xGrid^2,
                                                              Level3 = xGrid^3,
                                                              Level4 = xGrid^4))),
              colour="blue") +
    ggtitle("Truth or Bluff (Linear Regression)") +
    xlab("position level") +
    ylab("salary")

File: part_2/svr.py
import warnings
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR

warnings.filterwarnings(action="ignore", module="scipy", message="^internal gelsd")

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = np.ravel(sc_y.fit_transform(y.reshape(-1, 1)))
y.reshape(-1, 1)

regressor = SVR(kernel="rbf")
regressor.fit(X, y)

# predicting a new result with Polynomial Regression
y_pred = sc_y.inverse_transform(regressor.predict(sc_X.transform(np.array([[6.5]]))))
print(y_pred)

# Visualising the SVR results (for higher resolution and smoother curve)
# choice of 0.01 instead of 0.1 step because the data is feature scaled
X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape((len(X_grid), 1))

plt.scatter(X, y, color='red')
plt.plot(X_grid, regressor.predict(X_grid), color='blue')
plt.title('Truth or Bluff (SVR)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()
